File: HelloEarthSource/HelloEarth/CartoDBLayer.m
//
//  CartoDBLayer.m
//  HelloEarth
//
//  Created by Chris Lamb on 6/28/15.
//  Copyright (c) 2015 com.SantaCruzNewspaperTaxi. All rights reserved.
//

#import "CartoDBLayer.h"

@implementation CartoDBLayer

- (id)initWithSearch:(NSString *)inSearch
{
    self = [super init];
    search = inSearch;
    opQueue = [[NSOperationQueue alloc] init];
    
    return self;
}

- (void)startFetchForTile:(MaplyTileID)tileID forLayer:(MaplyQuadPagingLayer *)layer
{
    // bounding box for tile
    MaplyBoundingBox bbox;
    [layer geoBoundsforTile:tileID ll:&bbox.ll ur:&bbox.ur];
    NSURLRequest *urlReq = [self constructRequest:bbox];
    
    // kick off the query asychronously
    [NSURLConnection
     sendAsynchronousRequest:urlReq
     queue:opQueue
     completionHandler:
     ^(NSURLResponse *response, NSData *data, NSError *connectionError)
     {
         // parse the resulting GeoJSON
         MaplyVectorObject *vecObj = [MaplyVectorObject VectorObjectFromGeoJSON:data];
         if (vecObj)
         {
             // display a transparent filled polygon
             MaplyComponentObject *filledObj =
             [layer.viewC
              addVectors:@[vecObj]
              desc:@{kMaplyColor: [UIColor colorWithRed:0.25
                                                  green:0.0 blue:0.0 alpha:0.25],
                     kMaplyFilled: @(YES),
                     kMaplyEnable: @(NO)
                     }
              mode:MaplyThreadCurrent];
             
             // display a line around the lot
             MaplyComponentObject *outlineObj =
             [layer.viewC
              addVectors:@[vecObj]
              desc:@{kMaplyColor: [UIColor redColor],
                     kMaplyFilled: @(NO),
                     kMaplyEnable: @(NO)
                     }
              mode:MaplyThreadCurrent];
             
             // keep track of it in the layer
             [layer addData:@[filledObj,outlineObj] forTile:tileID];
         }
         
         // let the layer know the tile is done
         [layer tileDidLoad:tileID];
     }];
}

- (NSURLRequest *)constructRequest:(MaplyBoundingBox)bbox
{
    double toDeg = 180/M_PI;
    NSString *query = [NSString stringWithFormat:search,bbox.ll.x*toDeg,bbox.ll.y*toDeg,bbox.ur.x*toDeg,bbox.ur.y*toDeg];
    NSString *encodeQuery = [query stringByAddingPercentEscapesUsingEncoding:NSUTF8StringEncoding];
    encodeQuery = [encodeQuery stringByReplacingOccurrencesOfString:@"&" withString:@"%26"];
    NSString *fullUrl = [NSString stringWithFormat:@"https://pluto.cartodb.com/api/v2/sql?format=GeoJSON&q=%@",encodeQuery];
    NSURLRequest *urlReq = [NSURLRequest requestWithURL:[NSURL URLWithString:fullUrl]];
    
    return urlReq;
}

@end

File: tutorial/code/CartoDBLayer.swift
//
//  CartoDBLayer.swift
//  HelloEarth
//
//  Created by jmnavarro on 17/08/15.
//  Copyright (c) 2015 Mousebird. All rights reserved.
//

import Foundation

class CartoDBLayer: NSObject, MaplyPagingDelegate {

	private var _minZoom = Int32(0)
	private var _maxZoom = Int32(0)

	private var search: String
	private var opQueue: NSOperationQueue?

	// Create with the search string we'll use
	init(search: String) {
		self.search = search
		self.opQueue = NSOperationQueue()

		super.init()
	}

	func minZoom() -> Int32 {
		return _minZoom
	}

	func maxZoom() -> Int32 {
		return _maxZoom
	}

	func setMinZoom(value: Int32) {
		_minZoom = value
	}

	func setMaxZoom(value: Int32) {
		_maxZoom = value
	}

	func startFetchForTile(tileID: MaplyTileID, forLayer layer: MaplyQuadPagingLayer) {
		// bounding box for tile
		let bbox = layer.boundsForTile(tileID)
		let urlReq = constructRequest(bbox)

		let session = NSURLSession(configuration: NSURLSessionConfiguration.defaultSessionConfiguration())
		session.dataTaskWithRequest(urlReq) { (data, response, error) in
			// parse the resulting GeoJSON
			let str = NSString(data: data!, encoding: NSUTF8StringEncoding)
			print(str)
			if let data = data,
				vecObj = MaplyVectorObject(fromGeoJSON: data) {
				// display a transparent filled polygon
				let filledObj = layer.viewC!.addVectors([vecObj],
				                                        desc: [
															kMaplyColor: UIColor(red: 0.25, green: 0.0, blue: 0.0, alpha: 0.25),
															kMaplyFilled: true,
															kMaplyEnable: false],
				                                        mode: .Current)

				// display a line around the lot
				let outlineObj = layer.viewC!.addVectors([vecObj],
				                                         desc: [
															kMaplyColor: UIColor.redColor(),
															kMaplyFilled: false,
															kMaplyEnable: false],
				                                         mode: .Current)

				// keep track of it in the layer
				layer.addData([filledObj!, outlineObj!], forTile: tileID)

				// let the layer know the tile is done
				layer.tileDidLoad(tileID)
			}
			}.resume()
	}

	func constructRequest(bbox: MaplyBoundingBox) -> NSURLRequest {
		let toDeg = Float(180.0/M_PI)
		let query = NSString(format: search, bbox.ll.x * toDeg, bbox.ll.y * toDeg,bbox.ur.x * toDeg, bbox.ur.y * toDeg)
		var encodeQuery = query.stringByAddingPercentEncodingWithAllowedCharacters(NSCharacterSet.URLQueryAllowedCharacterSet())
		let range = encodeQuery!.startIndex..<encodeQuery!.endIndex
		encodeQuery = encodeQuery!.stringByReplacingOccurrencesOfString("&", withString: "%26", options: [], range: range)
		let fullUrl = NSString(format: "https://pluto.cartodb.com/api/v2/sql?format=GeoJSON&q=%@", encodeQuery!) as String
		return NSURLRequest(URL: NSURL(string: fullUrl)!)
	}

}
